service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/CbsSessionHandler.java

```diff
@@ -0,0 +1,193 @@
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.util.UUID;
+
+@Slf4j
+public class CbsSessionHandler extends ErrorLoggingBaseHandlerWithCleanup implements AuthenticationMessageCallback, LinkStateCallback
+{
+    private Session session;
+    private CbsSenderLinkHandler cbsSenderLinkHandler;
+    private CbsReceiverLinkHandler cbsReceiverLinkHandler;
+    private CbsSessionStateCallback cbsSessionStateCallback;
+    private final TokenCredential authenticationTokenProvider;
+    private final TokenCredentialType authorizationType;
+    private boolean senderLinkOpened = false;
+    private boolean receiverLinkOpened = false;
+
+    CbsSessionHandler(Session session, CbsSessionStateCallback cbsSessionStateCallback, TokenCredential authenticationTokenProvider, TokenCredentialType authorizationType)
+    {
+        this.session = session;
+
+        //All events in this reactor that happened to this session will be handled in this instance (onSessionRemoteOpen, for instance)
+        BaseHandler.setHandler(this.session, this);
+
+        this.cbsSessionStateCallback = cbsSessionStateCallback;
+        this.authenticationTokenProvider = authenticationTokenProvider;
+        this.authorizationType = authorizationType;
+        this.session.open();
+    }
+
+    @Override
+    public void onSessionLocalOpen(Event event)
+    {
+        this.session = event.getSession();
+
+        Sender cbsSender = this.session.sender(CbsSenderLinkHandler.getCbsTag());
+        this.cbsSenderLinkHandler = new CbsSenderLinkHandler(cbsSender, this, this.authenticationTokenProvider, this.authorizationType);
+
+        Receiver cbsReceiver = this.session.receiver(CbsReceiverLinkHandler.getCbsTag());
+        this.cbsReceiverLinkHandler = new CbsReceiverLinkHandler(cbsReceiver, this, this);
+    }
+
+    @Override
+    public void onSessionRemoteOpen(Event e)
+    {
+        log.trace("CBS session opened remotely");
+    }
+
+    @Override
+    public void onSessionLocalClose(Event e)
+    {
+        log.trace("CBS session closed remotely");
+        this.session.getConnection().close();
+
+        this.cbsSenderLinkHandler.close();
+        this.cbsReceiverLinkHandler.close();
+    }
+
+    @Override
+    public void onSessionRemoteClose(Event e)
+    {
+        Session session = e.getSession();
+        if (session.getLocalState() == EndpointState.ACTIVE)
+        {
+            this.close();
+        }
+    }
+
+    public void close()
+    {
+        log.trace("Closing this CBS session");
+        this.session.close();
+    }
+
+    @Override
+    public DeliveryState handleAuthenticationResponseMessage(int status, String description)
+    {
+        if (status == 200)
+        {
+            log.debug("CBS session successfully authenticated");
+            this.cbsSessionStateCallback.onAuthenticationSucceeded();
+        }
+        else
+        {
+            IOException e = new IOException(status + " : " + description);
```

Comment:
Why IOException? Send IotHubException?

Reply:
Yep, this should be an IotHubException, good catch